lms/djangoapps/completion/models.py

```diff
@@ -85,17 +86,28 @@ def submit_completion(self, user, course_key, block_key, completion):
                 "block_key must be an instance of `opaque_keys.edx.keys.UsageKey`.  Got {}".format(type(block_key))
             )
 
-        obj, isnew = self.get_or_create(
-            user=user,
-            course_key=course_key,
-            block_type=block_type,
-            block_key=block_key,
-            defaults={'completion': completion},
-        )
-        if not isnew and obj.completion != completion:
-            obj.completion = completion
-            obj.full_clean()
-            obj.save()
+        if waffle.waffle().is_enabled(waffle.ENABLE_COMPLETION_TRACKING):
+            obj, isnew = self.get_or_create(
+                user=user,
+                course_key=course_key,
+                block_type=block_type,
+                block_key=block_key,
+                defaults={'completion': completion},
+            )
+            if not isnew and obj.completion != completion:
+                obj.completion = completion
+                obj.full_clean()
+                obj.save()
+        else:
+            # If the featureis not enabled, return an empty, unsaved BlockCompletion object.
+            obj = self.model(
+                user=user,
+                course_key=course_key,
+                block_type=block_type,
+                block_key=block_key,
+                completion=0.0,
```

Comment:
- Could someone save this and accidentally create a record?  I don't know enough about Django.  Does that expose some risk?
- Are we thinking it is too painful to raise an exception here and keep all the calls from actually being made?

If we do stick with this type of implementation, would "isnew = False" below be better?

Reply:
@jcdyer: I'm responding to [your comment](https://github.com/edx/edx-platform/pull/16112#issuecomment-338743562) here.  

Basically, I want to ensure that when the code is disabled: 1) we don't write any data, and 2) we don't break active functionality.  I don't think there is a best way, so I just want to lay out my thoughts on the options, and you can decide if you want to stick with this implementation or pivot.

It seems like there are 3 options:
1. Raise an error.
2. Return a value that would make other code happy if it were using the return value (current implementation), or
3. Return a value like None that might break code looking for a proper tuple, but wouldn't cause an error otherwise.

If you raise an error here, it would ensure the code isn't doing anything we don't want it to do, but it "might" require extra code to avoid certain calls.  This seems the safest, but it depends how much additional code/conditionals you need, because each additional bit of code adds more room for error.

The current implementation, where we return a legitimate value, seems the most deceptive because it may make a caller happy, but we don't know who that caller is and whether or not we wish that caller to be happy.

The third is to return a value like None.  This is sort of a compromise that has some benefits and disadvantages of each of the earlier methods.  It also means no one could accidentally do anything with the return valie.

Let me know what you think.  Ignore my earlier comment about "isnew".  My wish is that this value would be totally unimportant in this case, otherwise, we might have a bug.